### PR TITLE
multiple-pipeline: simplify testing options with playback/capture/all 

### PR DESCRIPTION
**Pull request by fredoh9**
-f option was introduced to decide whether to fill the pipeline with
playback or capture first. But it causes more confusion. Now still keep
-f option, but to simplify the pipeline mode option.

playback mode : only plyaback pipeline
capture mode : only capture pipelines
all mode : run both playback and capture pipelines

Simple usage example:

* playback pipelines only
multiple-pipeline-playback.sh # playback with default values
multiple-pipeline-playback.sh -c 5 -l 15
or
multiple-pipeline.sh -f p -c 5 -l 15

* capture pipelines only
multiple-pipeline-capture.sh # capture with default values
multiple-pipeline-capture.sh -c 5 -l 15
or
multiple-pipeline.sh -f c -c 5 -l 15

* all pipelines, set pipeline count large enough (SOF CI need to change the command from '-f p' to '-f a')
multiple-pipeline.sh -f a -c 20 -l 50

fixes:
- https://github.com/thesofproject/sof-test/issues/1022

### CODE REVIEW

File: test-case/multiple-pipeline.sh [hunk 1]
```diff
@@ -71,12 +65,18 @@ logger_disabled || func_lib_start_log_collect
...
-# in topology is less than the required pipeline count, all pipeline will be started.
+if [ "$f_arg" == "p" ]; then
+    func_pipeline_export "$tplg" "type:playback & ${MULTI_PIPELINE_FILTER}"
+elif [ "$f_arg" == "c" ]; then
+    func_pipeline_export "$tplg" "type:capture & ${MULTI_PIPELINE_FILTER}"
+elif [ "$f_arg" == "a" ]; then
+    func_pipeline_export "$tplg" "type:any & ${MULTI_PIPELINE_FILTER}"
+else
+    die "not supported option $f_arg"
+fi
```
Comment (marc-hb) [inline test-case/multiple-pipeline.sh:76]:
```shell
case "$f_arg" in
   'p') .... ;;

    *) die "unsupported option $f_arg"
```

**Comment by fredoh9** [comment]
multiple-pipeline-all.sh from SOF CI is changed from '-f p' to '-f a'

**Review by marc-hb — COMMENTED** [review]
Missing this:
```diff
- OPT_DESC['f']='Fill either playback (p) or capture (c) first or any (a) for all pipelines'
+ OPT_DESC['f']='Fill either playback (p) or capture (c) or any (a) for all pipelines'
```

Can you please change the following line and add at least one `if/then/else`?
```
[[ $PIPELINE_COUNT -gt ${OPT_VAL['c']} ]] && max_count=${OPT_VAL['c']} || max_count=$PIPELINE_COUNT
```
Short-circuit with a mix of `&&` and `||` is really unreadable.

Do our test plans need some corresponding name changes? Maybe not, maybe they were vague enough :-)

File: test-case/multiple-pipeline.sh [hunk 2]
```diff
@@ -8,22 +8,16 @@
...
-##      3. all pipelines mode: run all pipelines in parallel by
-##          a. use option '-f a' OR
-##          b. specify a very big requested pipeline count to -c option
+##      1. Playback mode: playback pipelines ONLY are used, test pipeline count
+##         (OPT_VAL['c']) is respected, but maxed out with playback pipelines.
+##      2. Capture mode: same with playback mode, but capture pipelines ONLY will
+##         be used.
+##      3. All pipelines mode: Run all playback and capture pipelines in parallel
+##         by using the option '-f a' and set the number of pipelines with -c.
+##         If you need to use all pipelines, set a high pipeline count such as -c 20.
```
Comment (plbossart) [inline test-case/multiple-pipeline.sh:17]: what happens if the -c option is lower than the total of pipelines?

I also feel that this option could quickly run in trouble, with e.g. the nocodec topology we have multiple mixers and multiple copiers on capture, so we could end-up with 10+ streams in parallel. That's likely to fail in practice.

We may need to have two counts for capture and playback each, it's more pragmatic to me than trying to run 20 pipelines.

**Review by plbossart — COMMENTED** [review]
Thanks for taking the lead on this @fredoh9, much appreciated. See a couple of comments below, I am not sure having a large number of pipelines is realistic/feasible.

File: test-case/multiple-pipeline.sh (hunk 2 above)
Comment (fredoh9) [reply test-case/multiple-pipeline.sh:17]: If -c option has less than total pipeline. first fill with playback pipelines, rest of number will be filled with capture. so capture pipeline will be limited by the number.

For example, with -c 20, MTLP NOCODEC tplg has playback 4 pcm + capture 7 pcm = total 11 pcm. so mulitple-pipeline-all will start all 11 pcm at same time. If -c 8 is set, playback 4 pcm + capture 4 pcm will be used. If any pcm need to be skipped, there is a way to filter out by pcm name.

If we have two counts, there needs to be a little more condition checking between two counts. still trivial but I need more requirement for all(playback + capture) case.

**Comment by fredoh9** [comment]
> Missing this:
> 
> ```diff
> - OPT_DESC['f']='Fill either playback (p) or capture (c) first or any (a) for all pipelines'
> + OPT_DESC['f']='Fill either playback (p) or capture (c) or any (a) for all pipelines'
> ```
> 
> Can you please change the following line and add at least one `if/then/else`?
> 
> ```
> [[ $PIPELINE_COUNT -gt ${OPT_VAL['c']} ]] && max_count=${OPT_VAL['c']} || max_count=$PIPELINE_COUNT
> ```

Addressed above comments from @marc-hb.
I rebased top of main unnecessarily. diff from the previous commit has a bit more.

**Review by paulstelian97 — DISMISSED** [review]

File: test-case/multiple-pipeline.sh [hunk 3]
```diff
@@ -70,13 +64,22 @@ logger_disabled || func_lib_start_log_collect
...
+    *) die "not supported option $f_arg";;
+esac
+func_pipeline_export "$tplg" "$type_filter & ${MULTI_PIPELINE_FILTER}"
+
+# respect number of pipelines requested but max will be real pipeline count
+if [[ $PIPELINE_COUNT -gt ${OPT_VAL['c']} ]]; then
+    max_count=${OPT_VAL['c']}
+else
+    max_count=$PIPELINE_COUNT
+fi
```
Comment (marc-hb) [inline test-case/multiple-pipeline.sh:82]: maybe this:
```shell
min() { printf '%d' $(( "$1" < "$2"  ? "$1" : "$2" )); }

max_count=$(min "${OPT_VAL['c']}" $PIPELINE_COUNT)
```

**Review by marc-hb — COMMENTED** [review]
codestyle looks good, someone else must approve the logical changes.

For the record, missing blobs in
https://sof-ci.01.org/softestpr/PR1024/build295/devicetest/index.html
https://sof-ci.01.org/softestpr/PR1024/build296/devicetest/index.html
https://sof-ci.01.org/softestpr/PR1024/build297/devicetest/index.html

EDIT: fixed by Greg, it was some ssh key issue.

**Comment by marc-hb** [comment]
If you want you can fix the single shellcheck warning in https://github.com/thesofproject/sof-test/actions/runs/4746518584/jobs/8430264948?pr=1024 with this small change, preferably in a separate commit:

```diff
--- a/test-case/multiple-pipeline.sh
+++ b/test-case/multiple-pipeline.sh
@@ -102,7 +102,7 @@ func_run_pipeline_with_type()
         # convert array to line, shuf to get random line, covert line to array
         idx_lst=("$(seq 0 $((PIPELINE_COUNT - 1)) | sed 's/ /\n/g' | shuf | xargs)")
     fi
-    for idx in ${idx_lst[*]}
+    for idx in "${idx_lst[@]}"
     do
         channel=$(func_pipeline_parse_value "$idx" channel)
         rate=$(func_pipeline_parse_value "$idx" rate)

```

**Comment by fredoh9** [comment]
@marc-hb when I put double quote, I have expansion/substitution error. lost now

```
In test-case/multiple-pipeline.sh line 105:
    for idx in ${idx_lst[@]}
               ^-----------^ SC2068 (error): Double quote array expansions to avoid re-splitting elements.
```

Error message,
```
/home/ubuntu/sof-test/case-lib/pipeline.sh: line 81: [[: 0
                                                                                          1
                                                                                            2
                                                                                             3: syntax error in expression (error token is "1
    2
     3")
 /home/ubuntu/sof-test/case-lib/pipeline.sh: line 86: ${PIPELINE_0
                                                                         1
                                                                          2
                                                                           3[channel]}: bad substitution
```

**Review by marc-hb — COMMENTED** [review]
Everything looks great to me, @plbossart can you please take a look and confirm it addresses the bug you filed as you wanted it?

**Review by plbossart — APPROVED** [review]

**Review by marc-hb — APPROVED** [review]

**Pull request merged**